File: Data Structure/Queue/LNT_no1166_Recommended Results are Scattered.cpp
class Solution {
public:
    /**
     * @param elements: A list of recommended elements.
     * @param n: [picture P] can appear at most 1 in every n
     * @return: Return the scattered result.
     */
    vector<string> scatter(vector<string> &elements, int n) {
        // write your code here
        vector<string> res;
        int i = 0, size = elements.size();

        while (elements[i][0] != 'P') {
            res.push_back(elements[i]);
            i++;
        }

        queue<string> p, v;
        while (i < size) {
            if (elements[i][0] == 'P')
                p.push(elements[i]);
            else
                v.push(elements[i]);
            i++;
        }

        for (int j = 0;; j++) {
            string cur;
            if (j % n == 0 && !p.empty()) {
                cur = p.front();
                p.pop();
                res.push_back(cur);
            }
            else if (j % n != 0 && !v.empty()) {
                cur = v.front();
                v.pop();
                res.push_back(cur);
            }
            else
                break;
        }

        return res;
    }
};

File: Data Structure/Queue/LNT_no1166_Recommended Results are Scattered.py
class Solution:
    """
    @param elements: A list of recommended elements.
    @param n: [picture P] can appear at most 1 in every n
    @return: Return the scattered result.
    """
    def scatter(self, elements, n):
        # write your code here
        first_p = False
        p = collections.deque([])
        v = collections.deque([])
        res = []
        for s in elements:
            if not first_p:
                if s[0] == 'P':
                    first_p = True
                    p.append(s)
                else:
                    res.append(s)
            else:
                if s[0] == 'P':
                    p.append(s)
                else:
                    v.append(s)
        
        for i in range(len(p) + len(v)):
            if i % n == 0 and p:
                res.append(p.popleft())
            elif i % n != 0 and v:
                res.append(v.popleft())
            else:
                break
        
        return res
